packages/yoastseo/src/languageProcessing/researches/keywordCount.js

```diff
@@ -1,65 +1,100 @@
+import getMarkingsInSentence from "../helpers/highlighting/getMarkingsInSentence";
+import matchWordFormsWithSentence from "../helpers/match/matchWordFormsWithSentence";
+import isDoubleQuoted from "../helpers/match/isDoubleQuoted";
+import { markWordsInASentence } from "../helpers/word/markWordsInSentences";
 
 /**
- * Calculates the keyword count, takes morphology into account.
+ * Counts the occurrences of the keyphrase in the text and creates the Mark objects for the matches.
  *
- * @param {object} paper        The paper containing keyword and text.
- * @param {object} researcher   The researcher.
+ * @param {Sentence[]} sentences The sentences to check.
+ * @param {Array} topicForms The keyphrase forms.
+ * @param {string} locale The locale used in the analysis.
+ * @param {function} matchWordCustomHelper  A custom helper to match words with a text.
+ * @param {boolean} isExactMatchRequested Whether the exact matching is requested.
  *
- * @returns {object} An array of all the matches, markings and the keyword count.
+ * @returns {{markings: Mark[], count: number}} The number of keyphrase occurrences in the text and the Mark objects of the matches.
  */
-export default function( paper, researcher ) {
-	const topicForms = researcher.getResearch( "morphology" );
-	const memoizedTokenizer = researcher.getHelper( "memoizedTokenizer" );
-
-	// A helper to return all the matches for the keyphrase.
-	const matchWordCustomHelper = researcher.getHelper( "matchWordCustomHelper" );
-
-	let text = paper.getText();
-	text = removeHtmlBlocks( text );
-	const locale = paper.getLocale();
-
-	const sentences = getSentences( text, memoizedTokenizer );
-
-	const keywordsFound = {
-		count: 0,
-		matches: [],
-		sentencesWithKeywords: [],
-	};
+export function countKeyphraseInText( sentences, topicForms, locale, matchWordCustomHelper, isExactMatchRequested ) {
+	const result = { count: 0, markings: [] };
 
-	/*
-	 * Count the amount of keyphrase occurrences in the sentences.
-	 * An occurrence is counted when all keywords of the keyphrase are contained within the sentence. Each sentence can contain multiple keyphrases.
-	 * (e.g. "The apple potato is an apple and a potato." has two occurrences of the keyphrase "apple potato").
-	 */
 	sentences.forEach( sentence => {
-		const matchesInSentence = topicForms.keyphraseForms.map( keywordForms => matchWords( sentence,
-			keywordForms, locale, matchWordCustomHelper ) );
+		const matchesInSentence = topicForms.keyphraseForms.map( wordForms => matchWordFormsWithSentence( sentence,
+			wordForms, locale, matchWordCustomHelper, isExactMatchRequested ) );
 
-		const hasAllKeywords = matchesInSentence.every( keywordForm => keywordForm.count > 0 );
+		const hasAllKeywords = matchesInSentence.every( wordForms => wordForms.count > 0 );
 
 		if ( hasAllKeywords ) {
 			const counts = matchesInSentence.map( match => match.count );
+			const totalMatchCount = Math.min( ...counts );
```

Comment:
The number of the full-match count is the lowest count of the occurrences.
For example: 
* Keyphrase: red panda
* Keyphrase forms: [ [ "red" ], [ "panda", "pandas" ] ]
* Sentence: "The red pandas are relative to the giant pandas."
* Matches: [ [ {matches: ["red"], count: 1} ], [ {matches: ["pandas"], count: 2} ] ]
* Counts: [ 1, 2 ]

From the example above, the full-match is 1, because one of the "pandas" occurrences is not accompanied by "red" to be counted as a full-match.